plugins/elantp/fu-elantp-hid-haptic-device.c

```diff
@@ -0,0 +1,865 @@
+	else
+		fu_device_remove_flag(FU_DEVICE(parent), FWUPD_DEVICE_FLAG_IS_BOOTLOADER);
+
+	return TRUE;
+}
+
+static gint16
+fu_elantp_hid_haptic_device_ensure_eeprom_iap_ctrl(FuDevice *parent, FuElantpHidHapticDevice *self, GError **error)
+{
+	guint8 buf[2] = {0x0};
+	if (!fu_elantp_hid_haptic_device_read_cmd(parent, 
+						ETP_CMD_I2C_SET_EEPROM_CTRL, 
+						buf, 
+						sizeof(buf), 
+						error)) {
+		g_prefix_error(error, "failed to read IAPControl: ");
+		return -1;
+	}
+	self->iap_ctrl = fu_memread_uint16(buf, G_LITTLE_ENDIAN);
+
+	if ((self->iap_ctrl & 0x800) != 0x800) {
+        	g_prefix_error(error, "bit11 fail: ");
+		return -2;
+    	}
+    	if ((self->iap_ctrl & 0x1000) == 0x1000) {
+       		g_prefix_error(error, "bit12 fail resend: ");
+		return 0;
+    	}
+    
+	return 1;
+}
+
+static gboolean
+fu_elantp_hid_haptic_device_get_hatpic_driver_ic(FuDevice *parent, FuElantpHidHapticDevice *self, GError **error)
+{
+	guint8 buf[2] = {0x0};
+	guint16 value;
+	if (!fu_elantp_hid_haptic_device_read_cmd(parent, 
+						ETP_CMD_I2C_FLIM_TYPE_ENABLE, 
+						buf, 
+						sizeof(buf), 
+						error)) {
+		g_prefix_error(error, "failed to read haptic enable cmd: ");
+		return FALSE;
+	}
+	value = fu_memread_uint16(buf, G_LITTLE_ENDIAN);
+	
+	if (value == 0xFFFF || value == ETP_CMD_I2C_FLIM_TYPE_ENABLE) {
+		g_prefix_error(error, "failed to read haptic enable cmd: ");
+		return FALSE;
+	}
+
+	if (buf[0] & ETP_FW_FLIM_TYPE_ENABLE_BIT && buf[0] & ETP_FW_EEPROM_ENABLE_BIT) {
+		self->driver_ic = (buf[0] >> 4) & 0xF;
+		return TRUE;
+	}
+	
+	return FALSE;
+}
+
+static guint32
+fu_elantp_hid_haptic_device_get_version(FuDevice *parent, FuElantpHidHapticDevice *self, GError **error)
+{
+    	guint16 v_s = 0;
+    	guint16 v_d = 0;
+    	guint16 v_m = 0;
+    	guint16 v_y = 0;
+    	gint8 tmp[256] = {0x0};
+	guint8 buf[2] = {0x0};
+
+	if (!fu_elantp_hid_haptic_device_write_cmd(parent,
+						ETP_CMD_I2C_SET_EEPROM_CTRL,
+						ETP_CMD_I2C_GET_EEPROM_FW_VERSION,
+						error))
+		return 0;
+
+    	g_usleep(ELANTP_DELAY_RESET * 1000); 
+
+	if (!fu_elantp_hid_haptic_device_read_cmd(parent, 0x0321, buf, sizeof(buf), error)) {
+		g_prefix_error(error, "failed to read haptic version cmd: ");
+		return -1;
+	}
+    	v_d = buf[0];
+    	v_m = buf[1] & 0xF;
+   	v_s = (buf[1] & 0xF0) >> 4;
+
+	if (!fu_elantp_hid_haptic_device_write_cmd(parent,
+						ETP_CMD_I2C_SET_EEPROM_CTRL,
+						ETP_CMD_I2C_GET_EEPROM_IAP_VERSION,
+						error))
+		return 0;
+
+    	g_usleep(ELANTP_DELAY_RESET * 1000); 
+
+	if (!fu_elantp_hid_haptic_device_read_cmd(parent, 0x0321, buf, sizeof(buf), error)) {
+		g_prefix_error(error, "failed to read haptic version cmd: ");
+		return 0;
+	}
+    	v_y = buf[0];
+    	self->iap_ver = buf[1];
+
+    	if (v_y==0xFF || v_m==0xFF || v_d==0xFF || v_s==0xFF)
+		return 0;
+
+    	sprintf((char*)tmp, "%02d%02d%02d%02d", v_y, v_m, v_d, v_s);
+    	
+    	return atoi((char*)tmp);
```

Comment:
I don't think you need this as an integer, do fu_device_set_version() directly here